src/lib/comms/sol-lwm2m-bs-server.c

```diff
@@ -212,67 +213,95 @@ static const struct sol_coap_resource bootstrap_request_interface = {
+
+        switch (sec_modes[i]) {
+        case SOL_LWM2M_SECURITY_MODE_PRE_SHARED_KEY:
+            known_psks = va_arg(ap, struct sol_lwm2m_security_psk **);
+            SOL_NULL_CHECK_GOTO(known_psks, err_sec_modes);
+
+            cipher_suites[i] = SOL_SOCKET_DTLS_CIPHER_PSK_AES128_CCM8;
+            break;
+        case SOL_LWM2M_SECURITY_MODE_RAW_PUBLIC_KEY:
+            my_rpk = va_arg(ap, struct sol_lwm2m_security_rpk *);
+            SOL_NULL_CHECK_GOTO(my_rpk, err_sec_modes);
+            known_pub_keys = va_arg(ap, struct sol_blob **);
+            SOL_NULL_CHECK_GOTO(known_pub_keys, err_sec_modes);
+
+            cipher_suites[i] = SOL_SOCKET_DTLS_CIPHER_ECDHE_ECDSA_AES128_CCM8;
+            break;
+        case SOL_LWM2M_SECURITY_MODE_CERTIFICATE:
+            SOL_WRN("Certificate security mode is not supported yet.");
+            goto err_sec_modes;
+        case SOL_LWM2M_SECURITY_MODE_NO_SEC:
+            SOL_WRN("Bootstrap Server MUST use DTLS.");
+            goto err_sec_modes;
+        default:
+            SOL_WRN("Unknown DTLS Security Mode: %d", sec_modes[i]);
+            goto err_sec_modes;
+        }
     }
 
     server = calloc(1, sizeof(struct sol_lwm2m_bootstrap_server));
-    SOL_NULL_CHECK(server, NULL);
+    SOL_NULL_CHECK_GOTO(server, err_sec_modes);
 
     //LWM2M Bootstrap Server MUST always use DTLS
-    if (sec_mode == SOL_LWM2M_SECURITY_MODE_PRE_SHARED_KEY) {
-        sol_vector_init(&server->known_psks, sizeof(sol_lwm2m_security_psk));
-
-        SOL_VECTOR_FOREACH_IDX (known_psks, given_psk, i) {
-            psk = sol_vector_append(&server->known_psks);
-            SOL_NULL_CHECK_GOTO(psk, err_psk);
-            psk->id = sol_blob_ref(given_psk->id);
-            psk->key = sol_blob_ref(given_psk->key);
+    for (i = 0; i < num_sec_modes; i++) {
+        if (sec_modes[i] == SOL_LWM2M_SECURITY_MODE_PRE_SHARED_KEY) {
+            server->known_psks = known_psks;
+        } else if (sec_modes[i] == SOL_LWM2M_SECURITY_MODE_RAW_PUBLIC_KEY) {
+            server->rpk_pair.private_key = sol_blob_ref(my_rpk->private_key);
+            server->rpk_pair.public_key = sol_blob_ref(my_rpk->public_key);
+
+            server->known_pub_keys = known_pub_keys;
         }
     }
 
-    server->coap = sol_coap_server_new(&servaddr, true);
-    SOL_NULL_CHECK_GOTO(server->coap, err_coap);
-
-    server->security = sol_lwm2m_bootstrap_server_security_add(server, sec_mode);
-    if (!server->security) {
-        SOL_ERR("Could not enable %s security mode for LWM2M Bootstrap Server",
-            sec_mode_str);
-        goto err_security;
-    } else {
-        SOL_DBG("Using %s security mode", sec_mode_str);
+    server->coap = sol_coap_server_new_by_cipher_suites(&servaddr,
+        cipher_suites, num_sec_modes);
+    SOL_NULL_CHECK_GOTO(server->coap, err_rpk_pair);
+
+    for (i = 0; i < num_sec_modes; i++) {
+        server->security = sol_lwm2m_bootstrap_server_security_add(server, sec_modes[i]);
```

Comment:
No I'm not. Look at the code for `_security_add`: it works like an append operation. I add the sec_modes[i] to the same `server->security` object everytime - I only create the object if it doesn't exist already.